### PR TITLE
Make plugin hot-reloadable

### PR DESCRIPTION
**Pull request by intgr**
### Contributor checklist

- [x] I am targeting the `master` branch (and **not** the `release` branch)
- [x] I am using the provided [codeStyleConfig.xml](https://github.com/leinardi/mypy-pycharm/blob/release/.idea/codeStyles/codeStyleConfig.xml)
- [x] I have tested my contribution on these versions of PyCharm/IDEA:
 * PyCharm Community 2021.3
- [x] My contribution is fully baked and ready to be merged as is
- [x] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit 
      using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-using-keywords/)

----------

### Description
This allows the plugin to be (re)loaded without restarting the IDE, which is useful when developing the plugin and more convenient when upgrading.

There appear to be some issues in IntelliJ IDEA with dynamic unloading (IDEA-289241, IDEA-289243), but in the worst case, the user receives a message that they have to restart IDE to reload. Which was the norm previously.

Docs: https://plugins.jetbrains.com/docs/intellij/dynamic-plugins.html

### Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### CODE REVIEW

**Comment by intgr** [comment]
Hot-reloading sometimes works, but other times causes errors. Haven't figured out what is causing this.

<collapse>
java.lang.Throwable: Assertion failed
	at com.intellij.openapi.diagnostic.Logger.assertTrue(Logger.java:218)
	at com.intellij.openapi.diagnostic.Logger.assertTrue(Logger.java:227)
	at com.intellij.ide.plugins.PluginSet.enablePlugin(PluginSet.kt:48)
	at com.intellij.ide.plugins.DynamicPlugins.loadPlugin(DynamicPlugins.kt:820)
	at com.intellij.ide.plugins.DynamicPlugins.loadPlugin(DynamicPlugins.kt:808)
	at com.intellij.ide.plugins.DynamicPluginVfsListener$prepareChange$2$afterVfsChange$1.run(DynamicPluginVfsListener.kt:77)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:214)
	at com.intellij.openapi.application.TransactionGuardImpl.access$200(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:196)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$invokeLater$4(ApplicationImpl.java:348)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:82)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:131)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:47)
	at com.intellij.openapi.application.impl.FlushQueue$FlushNow.run(FlushQueue.java:187)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:891)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:760)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:492)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
</collapse>

**Comment by intgr** [comment]
Sigh. After spending hours on this (even before creating this PR), trying out lots of things, many dead ends, I'm finally closer to understanding this. Just writing down notes for myself for future reference.

* Helpful background for hot reloading plugin development: https://intellij-support.jetbrains.com/hc/en-us/community/posts/360010163299-How-to-Trigger-Dynamic-Plugin-Reload
* build.gradle:
  * `buildSearchableOptions.enabled = false` needs to be added (https://github.com/JetBrains/gradle-intellij-plugin/blob/master/FAQ.md#how-to-disable-building-searchable-options)
  * `systemProperties.put("ide.plugins.snapshot.on.unload.fail", "true")` in `runIde {}`
* gradle.properties: `org.gradle.jvmargs=-XX:+UnlockDiagnosticVMOptions`
* OK now I'm seeing errors in logs:
  ```
  Plugin com.leinardi.pycharm.mypy is not unload-safe because class loader cannot be unloaded.
  2022-02-19 17:26:49,160 [ 113120]   INFO - lij.ide.plugins.DynamicPlugins - Snapshot analysis result: Root 1:
  ROOT: Global JNI
     [...]
     com.intellij.openapi.util.Pair.first
     com.intellij.codeInsight.daemon.impl.HighlightInfo$IntentionActionDescriptor.myAction
     com.leinardi.pycharm.mypy.intentions.TypeIgnoreIntention.<class>          <-- smoking gun?
     com.leinardi.pycharm.mypy.intentions.TypeIgnoreIntention.<loader>         <-- smoking gun?
   * com.intellij.ide.plugins.cl.PluginClassLoader
  ```
* This may be helpful for debugging: https://intellij-support.jetbrains.com/hc/en-us/community/posts/360008211319-Plugin-not-unload-safe-because-class-loader-cannot-be-unloaded

**Comment by intgr** [comment]
Looks like hot-reload works as long as:
* The "Problems" tool window is closed.
* The editor cursor is not currently in a location with a Mypy inspection.

:facepalm: I'm glad this is cleared up now. Seems like something IntelliJ should fix upstream.

**Comment by intgr** [comment]
Reported bugs to IntelliJ: https://youtrack.jetbrains.com/issue/IDEA-289241 https://youtrack.jetbrains.com/issue/IDEA-289243

**Comment by leinardi** [comment]
Thanks a lot for the investigation and for opening the bug reports upstream! I voted and started watching both of them.

BTW, let me know when you think there are enough changes to publish a new release :+1:

**Comment by intgr** [comment]
I've tested this thoroughly enough that I think this can be merged.

The dynamic unloading issues in IntelliJ aren't blockers for this: in the worst case, the user receives a message that they have to restart the IDE to reload the plugin. Which was the norm previously.

When it does work, it saves a significant amount of time during development and updating.

File: build.gradle [hunk 1]
```diff
@@ -79,6 +79,13 @@ intellij {
 
 runIde {
     systemProperties.put("idea.log.debug.categories", "#com.leinardi.pycharm.mypy")
+    // Log verbose information when dynamic plugin unloading fails
+    systemProperties.put("ide.plugins.snapshot.on.unload.fail", "true")
+}
+
+// Causes error popup when building while sandbox IDE is open. Disable in development
+if (System.getenv('DEVELOP')) {
+    buildSearchableOptions.enabled = false
```
Comment (intgr) [inline build.gradle:88]: Do you have any better ideas how to disable this conditionally in development only? (When running the buildPlugin task)

Right now I changed IDE run configuration to set the "DEVELOP" environment variable. Works well enough for me.

Comment (leinardi) [reply build.gradle:88]: Not really, I think env variables are a good option in this case.

**Review by leinardi — APPROVED** [review]
> in the worst case, the user receives a message that they have to restart the IDE to reload the plugin. Which was the norm previously.

Oh good to know. I also saw that JetBrains assigned someone to the tickets. Let's merge it but I want to also test it before releasing it (can't do it right now).

**Comment by leinardi** [comment]
@intgr I tested locally your changes and everything seems to work fine for me! 
I just submitted a new release with all your changes: https://github.com/leinardi/mypy-pycharm/releases

Thanks a lot for all the help and effort you are putting in!

BTW, not sure if you noticed or if I already told you but this plugin code is heavily inspired by Checkstyle-IDEA: I basically forked that project and adapted it to run Mypy. If you are looking for inspiration or if you want to check if there are some interesting changes that we could backport I suggest to take a look at their commit log: https://github.com/jshiell/checkstyle-idea/commits/main

**Pull request merged**